### PR TITLE
Support dp test with txt

### PR DESCRIPTION
**Pull request by HuangJiameng**
Support using a text file(E.G. `list.txt`) to specify the list of test sets

list.txt
```
data_dir/sys.000
data_dir/sys.001
data_dir/sys.002
...
```
`/data_dir/sys.000` should be a directory with "type.raw".
Then one can use `-l list.txt` to replace `-s data_dir`.

### CODE REVIEW

File: deepmd/entrypoints/test.py [hunk 1]
```diff
@@ -57,9 +60,13 @@ def test(
     RuntimeError
         if no valid system was found
     """
-    all_sys = expand_sys_str(system)
-    if len(all_sys) == 0:
-        raise RuntimeError("Did not find valid system")
+    if(datafile is not None):
```
Comment (iProzd) [inline deepmd/entrypoints/test.py:63]: Bracket is not necessary

File: deepmd/entrypoints/test.py [hunk 2]
```diff
@@ -57,9 +60,13 @@ def test(
...
+        all_sys = open(datafile, 'r')
+        all_sys = [i.strip() for i in all_sys.readlines()]
+    else:
+        all_sys = expand_sys_str(system)
+        if len(all_sys) == 0:
```
Comment (iProzd) [inline deepmd/entrypoints/test.py:68]: Move this condition out of ‘if’ scope above.

File: deepmd/entrypoints/main.py [hunk 3]
```diff
@@ -271,6 +271,13 @@ def main_parser() -> argparse.ArgumentParser:
         type=str,
         help="The system dir. Recursively detect systems in this directory",
     )
+    parser_tst.add_argument(
+        "-f",
```
Comment (iProzd) [inline deepmd/entrypoints/main.py:275]: It seems that -f and -s can not exist at the same time. 
It's better to use 
'''
parser_tst_subgroup = parser_tst.add_mutually_exclusive_group()
parser_tst_subgroup.add_argument("-s", ... 
parser_tst_subgroup.add_argument("-f", ...
'''

**Review by iProzd — CHANGES_REQUESTED** [review]

File: deepmd/entrypoints/test.py [hunk 4]
```diff
@@ -57,9 +60,13 @@ def test(
     RuntimeError
         if no valid system was found
     """
-    all_sys = expand_sys_str(system)
-    if len(all_sys) == 0:
-        raise RuntimeError("Did not find valid system")
+    if(datafile is not None):
+        all_sys = open(datafile, 'r')
```
Comment (njzjz) [inline deepmd/entrypoints/test.py:64]: This file is not closed. See also https://stackoverflow.com/a/7396043/9567349

File: deepmd/entrypoints/test.py [hunk 5]
```diff
@@ -57,9 +60,13 @@ def test(
...
+        all_sys = [i.strip() for i in all_sys.readlines()]
```
Comment (njzjz) [inline deepmd/entrypoints/test.py:65]: Simpler:
```py
f.read().splitlines()
```

File: deepmd/entrypoints/test.py [hunk 6]
```diff
@@ -39,6 +40,8 @@ def test(
         path where model is stored
     system : str
         system directory
+    datafile: str
```
Comment (njzjz) [inline deepmd/entrypoints/test.py:43]: Numpydoc format: 
> The colon must be preceded by a space

```suggestion
    datafile : str
```

See https://numpydoc.readthedocs.io/en/latest/format.html

File: deepmd/entrypoints/main.py [hunk 7]
```diff
@@ -271,6 +271,13 @@ def main_parser() -> argparse.ArgumentParser:
...
+        "--datafile",
+        default=None,
+        type=str,
+        help="The list of systems to test.",
```
Comment (njzjz) [inline deepmd/entrypoints/main.py:279]: This help does not indicate that the input should be a filename.

**Review by njzjz — CHANGES_REQUESTED** [review]
Also, it is encouraged to add a test.

**Review by iProzd — APPROVED** [review]

File: deepmd/entrypoints/test.py [hunk 8]
```diff
@@ -57,7 +60,13 @@ def test(
     RuntimeError
         if no valid system was found
     """
-    all_sys = expand_sys_str(system)
+    if datafile is not None:
+        all_sys = open(datafile, 'r')
+        all_sys = all_sys.read().splitlines()
+        datafile.close()
```
Comment (njzjz) [inline deepmd/entrypoints/test.py:66]: `datafile` is `str`.

Comment (njzjz) [reply deepmd/entrypoints/test.py:66]:
```suggestion
        with open(datafile, 'r') as f:
            all_sys = f.read().splitlines()
```

**Review by njzjz — CHANGES_REQUESTED** [review]

File: deepmd/entrypoints/test.py (hunk 8 above)
Comment (njzjz) [reply deepmd/entrypoints/test.py:66]: By the way, you should manually test your commit if there is no unit test.

Comment (HuangJiameng) [reply deepmd/entrypoints/test.py]: done

**Review by njzjz — APPROVED** [review]

**Pull request merged**